Update servicing checklist. (#2857)
dotnet/source-build: .github/ISSUE_TEMPLATE/releases/release_checklist.md

<!--
  To start the checklist for a new release:
  - If it's an internal release, open a new issue in dotnet/release (private repo).
    - Otherwise, open a new issue in dotnet/source-build.
  - Paste a copy of the below checklist.
  - Delete lines starting with [Internal] if running a non-internal release.

>   - Delete lines starting with [Non-Internal] if running an internal release.

Reviewer: Are we going to have non-internal releases for 3.1?  If there are, to simplify this workflow, what would be wrong with treating all released as internal?  It would reduce the chances of errors.

<!--
  To start the checklist for a new release:
  - Open a new issue in dotnet/release (private repo):
      - Issue title: Source-build {runtime-version} / {sdk-version} {Month} {Year} Servicing
  - Paste a copy of the below checklist.

-->

<!--
  Enter the version tracked by this checklist.
-->
# {runtime-version} / {sdk-version}

1. - [ ] Create a OneNote page in the NET Core Acquisition and Deployment
    OneNote > source-build > Servicing > Retrospective > "May 2022"
    (for example) to take ongoing notes on problems with the process,
    workarounds, and fixes.
      - This is useful to make sure context is available to review later.
        It may end up blank if the release goes very smoothly.
      - There are other notes on servicing in this OneNote. It may be
        useful to review if something goes wrong to see if it's been fixed
        before.
1. - [ ] Work in the [dotnet-source-build](https://dnceng.visualstudio.com/internal/_git/dotnet-source-build?path=%2F&version=GBmain&_a=contents)
         repo.
1. - [ ] Ensure [internal/release/3.1](https://dnceng.visualstudio.com/internal/_git/dotnet-source-build?path=/&version=GBinternal/release/3.1&_a=history)

>          branch is up to date with mirrored [release/3.1](https://github.com/dotnet/source-build/commits/release/3.1)

Reviewer: ```suggestion
         branch contains all commits made to the public [release/3.1](https://github.com/dotnet/source-build/commits/release/3.1) GitHub branch during this servicing cycle.
```

         branch.
1. - [ ] Update to new version.
      - [/Documentation/servicing/update-3.1.md](https://github.com/dotnet/source-build/tree/release/3.1/Documentation/servicing/update-3.1.md)

>       - Start a new working branch and PR for WIP so others can

Reviewer: ```suggestion
      - Start a new working branch off of [internal/release/3.1](https://dnceng.visualstudio.com/internal/_git/dotnet-source-build?path=/&version=GBinternal/release/3.1&_a=history) and open a PR for WIP so others can
```

>       - Start a new working branch off of [internal/release/3.1](https://dnceng.visualstudio.com/internal/_git/dotnet-source-build?path=/&version=GBinternal/release/3.1&_a=history) and open a PR for WIP so others can

Author: I think for this scenario it's useful to have a PR so we have an easy process to sign off on the changes.  Does that seem reasonable?

        - [ ] Add the servicing feed to build.sh and smoke-testNuGet.config
          1. The URL will look like this: <https://pkgs.dev.azure.com/dnceng/internal/_packaging/3.1.419-servicing.22219.28-shipping/nuget/v3/index.json>
          1. In build.sh, add this feed to the `VSS_NUGET_EXTERNAL_FEED_ENDPOINTS`.
          1. In smoke-testNuGet.config, add another feed entry that looks like
             the others, with key="5.0.408-servicing.22219.28-shipping"
             (for example) and value as the servicing url.
      - [ ] In `eng/Version.Details.Xml`, verify that the `<Sha>` tag for all
            repos were updated, except for these expected to usually be stable:
        - arcade
        - sourcelink
        - aspnet-xdt
        - cliCommandLineParser
        - xliff-tasks
        - Newtonsoft.Json
        - common
        - Roslyn (except for feature band updates)
        - MSBuild (except for feature band updates)
        - NuGet-client (except for feature band updates)
        - fsharp (except for feature band updates)
        - VSTest (except for feature band updates)
1. - [ ] [Regenerate patches as necessary](patch-updates.md).

>       - Run `./build.sh` to begin clone and patch application and see what,

Reviewer 2: This step will probably fail if we don't have the PAT setup from the steps below. Consider moving this after "Start building locally..." and before "Review prebuilt baseline diff...".

Author: Good point, changed as suggested.

        if anything, needs to be regenerated.
      - ASP.NET will always need to be regenerated because it contains a
        version number that needs to be updated.  The version should be updated
        to the SDK version being built.  See <https://github.com/dotnet/source-build/blob/release/3.1/patches/aspnetcore/0006-Fix-version-number.patch>.
1. - [ ] [Generate a PAT for AzDo](https://dev.azure.com/dnceng/_usersSettings/tokens).
      - This should include the `dnceng` organization.
      - Scope should include Read for Code and Packages.
1. - [ ] Start building locally (for quick diagnosis) and in CI (for super
         clean build) and iterate towards a green build.
      - For local builds:

>         - Set environment variable `internalPackageFeedPat` to the generated PAT.

Reviewer: ```suggestion
        - Set environment variable `internalPackageFeedPat` to the PAT generated in the previous step.
```

          - E.g. run `export internalPackageFeedPat;read internalPackageFeedPat`
            and paste the PAT. This sets up your running shell for an authenticated
            build. Passing `/p:` isn't sufficient.
        - Run `build.sh /p:AzDoPat=$internalPackageFeedPat`.
      - For CI builds:

>         - To run CI, from [dnceng/internal/source-build-pre-arcade-CI](https://dev.azure.com/dnceng/internal/_build?definitionId=251),

Reviewer: Consider adding a comment to ensure all local changes are pushed to the dnceng remote.

Author: Added.

          click "Run pipeline" and select your dev/… branch under "Branch/tag",
          then click "Run".
        - If you are unsure if your build will be successful, you can run only
          the centos71_* stages to save resources by selecting "Stages to run"
          in the run dialog.
1. - [ ] Review prebuilt baseline diff and iterate builds to drive to zero.
      - Prebuilts are packages used in a source-build that come from outside
        of source-build.  These are not allowed by our Linux partners so we
        have to eliminate them before shipping.
      - If there's a prebuilt issue in a build, it will fail with an error
        similar to:

>           ```

Reviewer: ```suggestion
          ```text
```

          a similar diff in the past.  The Source-Build Teams channel is a good
          place to start the conversation, or the dotnetADTeam DL.
1. - [ ] Add team members crummel, LoganBussell, and msimons as reviewers to
         the PR. Get acceptance.
      - Note: Don't merge the PR.  This is just used to generate the pre-embargo-date
        tarball for our partners.  This will be ported to the public repo
        later and become the official release.
1. - [ ] Download the artifacts from the centos71_Online_BaseTarball PR
         validation build onto a machine with Bash.
      - From the "Artifacts" page, navigate to "Tarball centos71_Online_BaseTarball"
        and download the `base_tarball_*.tar.gz` and `tarball_*-smoke-test-prereqs.tar.gz`
        files.
      - The smoke-test prereqs are platform-independent but should be taken
        from this non-portable build, *not* CentOS7 Offline Portable.
      - Can use <https://github.com/dotnet/source-build/blob/release/3.1/scripts/fetch-azdo-artifact-tarball.sh>
        as a library.
        - E.g. download the tarballs with `download_tarball`, and fix them up
          with `fix_azdo_tarball` if necessary.

> 1. - [ ] Rename tarball and smoke-test prereqs to human-friendly names.

Reviewer: Specify the expected human-friendly names.

Author: Added.

         (Can use `rename_tarball_inner_dir`.)
      - Refer to previous release's email threads for naming pattern.
1. - [ ] Upload tarballs and smoke-test prereqs to the `dotnetclimsrc` blob
         storage under the `source-build` container in `red-hat/3.1/` under a
         new folder named for the runtime and SDK version.  The access key for
         this blob storage is in the `EngKeyVault` vault in the "Dotnet
         Engineering Services" subscription.

> 1. - [ ] Complete manual validation steps

Reviewer: It feels like there should be a checkpoint where a source-build team member signs off.  Maybe that is concurrent w/the manual platform tests.  Having it earlier may be beneficial though just to give a once-over on the changes.  This would be especially beneficial on the first couple of runs.

Author: There's an earlier step:
> Add team members crummel, LoganBussell, and msimons as reviewers to the PR. Get acceptance.

on line 151.

>       - For 3.1, the build is poison-checked during the CI build.  If you get

Author: Cleaned this up a bit.  I still wanted to specify just in case someone ends up looking at the wrong one.

        an error related to "leak detection" or "CheckForPoison", get the source-build
        team involved as these can be difficult to diagnose and fix.
      - [ ] Verify the packs included in the SDK are from source-build-reference-packages,
        not source-build.
        - After smoke-tests have run, in `testing-smoke/builtCli`, run
          `strings sdk/<version>/Sdks/Microsoft.NET.Sdk/tools/net472/Microsoft.CSharp.dll
          | grep "built by"` and make sure this includes `built by: SOURCEBUILD`.

>       - [ ] Verify the packs included in the SDK have XML documentation comments

Reviewer: This step doesn't seem actionable.  How will the person running this step know if there is an issue or not?

Is this something a source-build team member should sign-off on?

Author: Added threshold for an error.  I don't know that we need team signoff on this one but I can add it if you feel the threshold is insufficient.

        - From the `testing-smoke/builtCli` directory, run `ls packs/Microsoft.NETCore.App.Ref/<version>/ref/netcoreapp3.1/*.xml`.
          There should be about 110 XML files here.
        - From the `testing-smoke/builtCli` directory, run `ls packs/Microsoft.AspNetCore.App.Ref/<version>/ref/netcoreapp3.1/*.xml`.
          There should be about 130 XML files here.
        - Issue tracking automation of this: [#1579](https://github.com/dotnet/source-build/issues/1579)
      - [ ] Re-validate the SHA1s in the `eng/Version.Details.xml` file with
        the manifest in VSU share dir.

>       - [ ] Review the file list diff between Microsoft-built SDK and source-built

Reviewer: Is this something a source-build team member should sign-off on?

Author: Added source-build team tag.

        SDK in smoke-test output.
        - For internal builds, smoke-test fails to acquire the Microsoft-built
          SDK. Manually acquire the Microsoft-built SDK from the build email
          you should have received, then run `scripts/compare-builds.sh`
          yourself to get the output to review.
          - [#1631](https://github.com/dotnet/source-build/issues/1631)
            tracks a fix.
        - Use <https://gist.github.com/omajid/9111a9310d452b3594b770d4ca8a3d87>
          as a loose baseline.
          - [#1630](https://github.com/dotnet/source-build/issues/1630) tracks
            a checked-in baseline.

> 1. - [ ] Complete manual smoke-testing on our primary testing platforms:

Reviewer: This should have steps for what is expected here - eg take the tarball, run prep, build, and smoke-tests.

Author: Added steps.

      - [ ] RHEL 7 VM
      - [ ] RHEL 8 VM
      - [ ] Fedora 30 - VM or [Docker](mcr.microsoft.com/dotnet-buildtools/prereqs:fedora-30-38e0f29-20191126135223)
      - [ ] CentOS 7 - VM or [Docker](mcr.microsoft.com/dotnet-buildtools/prereqs:centos-7-3e800f1-20190501005343)
      - [ ] CentOS 8 - VM or [Docker](mcr.microsoft.com/dotnet-buildtools/prereqs:centos-8-daa5116-20200325130212)
      - [ ] Debian 9 - VM or [Docker](mcr.microsoft.com/dotnet-buildtools/prereqs:debian-stretch-d61254f-20190807161114)
1. - [ ] [Source-build team] Send the tarball to partners. Include info about
     how certain we are that this will be the final Microsoft build.
      - Never overwrite a tarball. At least change the blob storage virtual
        dir to represent a new build. This can help avoid timing issues and
        make it more obvious if stale links were accidentally re-sent rather
        than new ones.
1. - [ ] SYNC POINT: Wait for Microsoft build release.
      - Wait for dotnet/announcements post. `Watch` the repo to get emails so
        you don't have to poll.
        - Example: <https://github.com/dotnet/announcements/issues/160>.

> 1. - [ ] In a dev branch on your GitHub fork of source-build, clean up.

Author: Added, thanks for the catch.

1. - [ ] Push the servicing branch to your GitHub fork and submit a
         source-build PR.
1. - [ ] If a build hits `fatal: reference is not a tree: cb5f173b`,
        [make the commit available on a fork and switch to it temporarily](https://github.com/dotnet/source-build/tree/release/3.1/Documentation/servicing/make-commit-available.md).
1. - [ ] Add crummel, MichaelSimons, and lbussell as reviewers.
1. - [ ] When CI is green and two reviewers approve, merge.
      - Avoid squash/rebase: nice to preserve commit hashes. However, there
        are no known dependencies on *source-build* commits being preserved.
1. - [ ] Create and push tags on the post-merge commit. [/Documentation/servicing/tagging.md](https://github.com/dotnet/source-build/tree/release/3.1/Documentation/servicing/tagging.md)
1. - [ ] Create a GitHub release on the tag with SDK versioning.
      - Go to <https://github.com/dotnet/source-build/tags>
      - Expand the SDK-versioned tag's message and copy it
      - Click `...` on the right on the SDK-versioned tag and press "Create release"
      - Paste into the release title
      - Click submit.

> 1. - [ ] [Source-build team] Notify distro maintainers and partners about

Reviewer: What channels are used for this?  Is there a template for the notification?

Author: Added Slack and GitHub and the recommended information.

  - Paste a copy of the below checklist.

-->

<!--
  Enter the version tracked by this checklist.
-->
# {runtime-version} / {sdk-version}

1. - [ ] Create a OneNote page in the NET Core Acquisition and Deployment
    OneNote > source-build > Servicing > Retrospective > "May 2022"
    (for example) to take ongoing notes on problems with the process,
    workarounds, and fixes.
      - This is useful to make sure context is available to review later.
        It may end up blank if the release goes very smoothly.
      - There are other notes on servicing in this OneNote. It may be
        useful to review if something goes wrong to see if it's been fixed
        before.
1. - [ ] Work in the [dotnet-source-build](https://dnceng.visualstudio.com/internal/_git/dotnet-source-build?path=%2F&version=GBmain&_a=contents)
         repo.
1. - [ ] Ensure [internal/release/3.1](https://dnceng.visualstudio.com/internal/_git/dotnet-source-build?path=/&version=GBinternal/release/3.1&_a=history)
         branch contains all commits made to the public [release/3.1](https://github.com/dotnet/source-build/commits/release/3.1) GitHub branch during this servicing cycle.

>          branch.

Author: Thanks, removed.

1. - [ ] Update to new version.

>       - [/Documentation/servicing/update-3.1.md](https://github.com/dotnet/source-build/tree/release/3.1/Documentation/servicing/update-3.1.md)

Author: Changed as suggested.

1. - [ ] Start building locally (for quick diagnosis) and in CI (for super
         clean build) and iterate towards a green build.
      - For local builds:
        - Set environment variable `internalPackageFeedPat` to the PAT generated in the previous step.
          - E.g. run `export internalPackageFeedPat;read internalPackageFeedPat`
            and paste the PAT. This sets up your running shell for an authenticated
            build. Passing `/p:` isn't sufficient.
        - Run `build.sh /p:AzDoPat=$internalPackageFeedPat`.
      - For CI builds:
        - Make sure all your changes are pushed to the dnceng internal remote.
          - Very importantly, *do not* push any changes to any GitHub remote.  
        - To run CI, from [dnceng/internal/source-build-pre-arcade-CI](https://dev.azure.com/dnceng/internal/_build?definitionId=251),
          click "Run pipeline" and select your dev/… branch under "Branch/tag",
          then click "Run".
        - If you are unsure if your build will be successful, you can run only
          the centos71_* stages to save resources by selecting "Stages to run"
          in the run dialog.

> 1. - [ ] [Regenerate patches as necessary](patch-updates.md).

Author: Updated in a way that I think makes this clear.  Let me know what you think.

      - [ ] Verify the packs included in the SDK are from source-build-reference-packages,
        not source-build.
        - After smoke-tests have run, in `testing-smoke/builtCli`, run
          `strings sdk/<version>/Sdks/Microsoft.NET.Sdk/tools/net472/Microsoft.CSharp.dll
          | grep "built by"` and make sure this includes `built by: SOURCEBUILD`.
      - [ ] Verify the packs included in the SDK have XML documentation comments
        - From the `testing-smoke/builtCli` directory, run `ls packs/Microsoft.NETCore.App.Ref/<version>/ref/netcoreapp3.1/*.xml`.
          There should be about 110 XML files here.  If there are fewer than 50, there is definitely a problem.
        - From the `testing-smoke/builtCli` directory, run `ls packs/Microsoft.AspNetCore.App.Ref/<version>/ref/netcoreapp3.1/*.xml`.
          There should be about 130 XML files here.  If there are fewer than 50, there is definitely a problem.
        - Issue tracking automation of this: [#1579](https://github.com/dotnet/source-build/issues/1579)
      - [ ] Re-validate the SHA1s in the `eng/Version.Details.xml` file with
        the manifest in VSU share dir.

>       - [ ] [Source-build team] Review the file list diff between Microsoft-built SDK and source-built

Author: Changed to reflect this.

          you should have received, then run `scripts/compare-builds.sh`
          yourself to get the output to review.
          - [#1631](https://github.com/dotnet/source-build/issues/1631)
            tracks a fix.
        - Use <https://gist.github.com/omajid/9111a9310d452b3594b770d4ca8a3d87>
          as a loose baseline.
          - [#1630](https://github.com/dotnet/source-build/issues/1630) tracks
            a checked-in baseline.
1. - [ ] Complete manual smoke-testing on our primary testing platforms:
      - Steps:
        1. Download the tarball on to each of the following platforms.
        2. Extract it to a drive with at least 150GB of empty space.
        3. Run `./prep.sh` to populate the previously-source-built and SBRP tarballs and the bootstrap SDK.
        4. Run `sudo unshare -n ./build.sh` to build the tarball.  Unshare is usually located in a package called `util-linux` or `moreutils`.  Or if running in Docker, use the `--offline` flag.
        5. If not running in Docker, use `sudo chown -R $(whoami) .` to retake ownership of all files.
        6. Run `./build.sh --run-smoke-test` to run the source-build tests.
      - [ ] RHEL 7 VM
      - [ ] RHEL 8 VM

>       - [ ] Fedora 30 - VM or [Docker](mcr.microsoft.com/dotnet-buildtools/prereqs:fedora-30-38e0f29-20191126135223)

Author: Thanks, fixed.

      - [ ] CentOS 7 - VM or [Docker](mcr.microsoft.com/dotnet-buildtools/prereqs:centos-7-3e800f1-20190501005343)
      - [ ] CentOS 8 - VM or [Docker](mcr.microsoft.com/dotnet-buildtools/prereqs:centos-8-daa5116-20200325130212)
      - [ ] Debian 9 - VM or [Docker](mcr.microsoft.com/dotnet-buildtools/prereqs:debian-stretch-d61254f-20190807161114)
1. - [ ] [Source-build team] Send the tarball to partners. Include info about
     how certain we are that this will be the final Microsoft build.

>       - Never overwrite a tarball. At least change the blob storage virtual

Author: Moved as suggested.

        you don't have to poll.
        - Example: <https://github.com/dotnet/announcements/issues/160>.
1. - [ ] In a dev branch on your GitHub fork of source-build, clean up.
      - [ ] Convert internal URIs to public in `eng/Version.Details.xml`.
        - `https://dev.azure.com/dnceng/internal/_git/<org>-<repo>` => `https://github.com/<org>/<repo>`
      - [ ] Convert internal `<SourceDirectory>` tags in `repo/*.proj` to the external ones. 
      - [ ] Remove internal package feeds from source-build.
        - `/NuGet.Config`
        - `/smoke-testNuGet.Config`
        - Do not remove from subrepos. The build infra removes these sources
          automatically while modifying each subrepo nuget.config file.
1. - [ ] Push the servicing branch to your GitHub fork and submit a
         source-build PR.
1. - [ ] If a build hits `fatal: reference is not a tree: cb5f173b`,
        [make the commit available on a fork and switch to it temporarily](https://github.com/dotnet/source-build/tree/release/3.1/Documentation/servicing/make-commit-available.md).
1. - [ ] Add crummel, MichaelSimons, and lbussell as reviewers.

> 1. - [ ] When CI is green and two reviewers approve, merge.

Author: Added some more notes to the `fatal: reference is not a tree` step.

          if anything, needs to be regenerated.
        - ASP.NET will always need to be regenerated because it contains a
          version number that needs to be updated.  The version should be updated
          to the SDK version being built.  See <https://github.com/dotnet/source-build/blob/release/3.1/patches/aspnetcore/0006-Fix-version-number.patch>.
        - Reach out to the source-build team if you are stuck on a patch issue.
      - For CI builds:
        - After patches have been updated, you can start running CI builds to make sure a clean build works.
        - Make sure all your changes are pushed to the dnceng internal remote.
          - Very importantly, *do not* push any changes to any GitHub remote.  
        - To run CI, from [dnceng/internal/source-build-pre-arcade-CI](https://dev.azure.com/dnceng/internal/_build?definitionId=251),
          click "Run pipeline" and select your dev/… branch under "Branch/tag",
          then click "Run".
        - If you are unsure if your build will be successful, you can run only
          the centos71_* stages to save resources by selecting "Stages to run"
          in the run dialog.

> 1. 

Reviewer 2: Typo: blank bullet point

Author: Thanks, fixed.

> -->

Author: I don't see it in the [6.0 checklist](https://raw.githubusercontent.com/dotnet/source-build/release/6.0/.github/ISSUE_TEMPLATE/releases/release_checklist.md), does GitHub take care of the title in that case?  Where does it need to go?  Or do you mean just adding it as the specified title for the issue?

> -->

Reviewer: It's a comment so you only see it in the un-rendered MD.

# {runtime-version} / {sdk-version}

> 1.  - [ ] Copy this template to a new issue in [dotnet/release](https://github.com/dotnet/release/issues) and fill in the runtime and SDK versions above.  Update this issue as you progress with servicing.

Author: Removed.

> 1.  - [ ] Create a OneNote page in [NET Core Acquisition and Deployment XXX Can we link to this?](https://microsoft.sharepoint.com/placeholderlink) > source-build > Servicing > Retrospective > "May 2022" (for example) to take ongoing notes on problems with the process, workarounds, and fixes.

Reviewer: Placeholder link should be removed.

Author: Removed.

1.  - [ ] [Internal] Work in the [dotnet-source-build](https://dnceng.visualstudio.com/internal/_git/dotnet-source-build?path=%2F&version=GBmain&_a=contents) repo.
    - [ ] [Non-Internal] Work in the [GitHub source-build](https://github.com/dotnet/source-build) repo.
1.  - [ ] [Internal] Ensure [internal/release/3.1](https://dnceng.visualstudio.com/internal/_git/dotnet-source-build?path=/&version=GBinternal/release/3.1&_a=history) branch is up to date with mirrored [release/3.1](https://github.com/dotnet/source-build/commits/release/3.1) branch.
1.  - [ ] Update to new version.
      - [3.1] [/Documentation/servicing/update-3.1.md](https://github.com/dotnet/source-build/tree/release/3.1/Documentation/servicing/update-3.1.md)
      - Start a new working branch and PR for WIP so others can easily look if needed and CI starts running.
        - [Internal] For internal builds, use the branch name `dev/<youralias>/<SDK-version>-<month name><YYYY>`.  This is necessary because some branch protection kicks in on branches with "release" in the name.
1.  - [ ] Version update cleanup and confirmations

>       - [3.1]

Reviewer: Since this is in the 3.1 branch, I think there 3.1 prefixes should be removed.  This one doesn't add value.

>         - [ ] Verify that the `PrivateSourceBuiltArtifactsPackageVersion` in `eng/Versions.props` matches N-1 release artifacts.

Reviewer: How do you determine what the N-1 release artifacts are?  I think this is going to need explanation.

Author: Added VaultExplorer and Storage Explorer instructions and details on how to find this version.

>         - [ ] Update the `global.json` SDK version to N-1: the latest released "3.1.XYY" SDK version, where "3.1.X" matches the SDK we're building.

Reviewer: This should be more explicit on what N-1 means.  An example would he helpful.  e.g. If the SDK release is 3.1.214, then N-1 would be 3.1.213.

        - [ ] Update feeds:

>              - [ ] If new feeds were added to NuGet.config, copy them to smoke-testNuGet.config and the `VSS_NUGET_EXTERNAL_FEED_ENDPOINTS` section in `build.sh`.

Reviewer: How does the reader know if new feeds were added?

Author: Added clarification and an example.

             - [ ] Add the servicing feed to build.sh and smoke-testNuGet.config
				1. The URL will look like this: https://pkgs.dev.azure.com/dnceng/internal/_packaging/3.1.419-servicing.22219.28-shipping/nuget/v3/index.json
				1. In build.sh, add this feed to the `VSS_NUGET_EXTERNAL_FEED_ENDPOINTS`.
				1. In smoke-testNuGet.config, add another feed entry that looks like the others, with key="5.0.408-servicing.22219.28-shipping" (for example) and value as the servicing url.

>         - [ ] Verify that all repos were updated, except for these expected to usually be stable:

Reviewer: What does it mean for a repo to be updated?

Reviewer 2: ```suggestion
        - [ ] In `eng/Version.Details.Xml`, verify that the `<Sha>` tag for all repos were updated, except for these expected to usually be stable:
```

Side note: do we typically see *all* repos updated in 3.1?

Author: No, there's a list of the usually-stable ones below:
          - arcade
          - sourcelink
          - aspnet-xdt
          - cliCommandLineParser
          - xliff-tasks
          - Newtonsoft.Json
          - common
          - Roslyn (except for feature band updates)
          - MSBuild (except for feature band updates)
          - NuGet-client (except for feature band updates)
          - fsharp (except for feature band updates)
          - VSTest (except for feature band updates)

          - NuGet-client (except for feature band updates)
          - fsharp (except for feature band updates)
          - VSTest (except for feature band updates)

> 1.  - [ ] Regenerate patches as necessary.

Reviewer: This should call out in more detail what it means to regenerate patches.

Author: Added a new doc with with details.

      - Run `./build.sh` to begin clone and patch application and see what, if anything, needs to be regenerated.  Issue to track re-patching workflow: [#1468](https://github.com/dotnet/source-build/issues/1468)
      - [3.1] ASP.NET will always need to be regenerated because it contains a version number that needs to be updated.  The version should be updated to the SDK version being built.  See https://github.com/dotnet/source-build/blob/release/3.1/patches/aspnetcore/0006-Fix-version-number.patch

> 1.  - [ ] Start building locally (for quick diagnosis) and in CI (for super clean build) and iterate towards a green build.

Reviewer: This step has no guidance for what to do if it is a public build.

Author: Removed public steps throughout.

>       - [Internal] [3.1] Set environment variable `internalPackageFeedPat` to dnceng internal PAT with package read and code read access.

Reviewer: This should include the steps on how to retrieve the PAT.

Author: Added instructions for generating a PAT.

>       - [Internal] [3.1] Set environment variable `internalPackageFeedPat` to dnceng internal PAT with package read and code read access.

Reviewer: What do you think about separating local builds from CI more.  The first two bullets are about local builds.  How about putting them under a common local build bullet?

Author: Changed as suggested.

        - E.g. run `export internalPackageFeedPat;read internalPackageFeedPat` and paste the PAT. This sets up your running shell for an authenticated build. Passing `/p:` isn't sufficient.

>       - [Internal] [3.1] Also pass arg `/p:AzDoPat=$internalPackageFeedPat` to build.sh.

Reviewer: ```suggestion
      - [Internal] [3.1] run `build.sh /p:AzDoPat=$internalPackageFeedPat`.
```

      - [Internal] To run CI, from [dnceng/internal/source-build-pre-arcade-CI](https://dev.azure.com/dnceng/internal/_build?definitionId=251), click "Run pipeline" and select your dev/… branch under "Branch/tag", then click "Run"

>          - If you are unsure if your build will be successful, you can run only the centos71_* stages to save resources.

Reviewer: How do you do this?

Author: Added details.

> 1.  - [ ] Review prebuilt baseline diff and iterate builds to drive to zero.

Reviewer: Where do you find the prebuilt report?  We should provide context as to what a prebuilt is and how you drive to zero.

Author: Added some details on this.

      - Find an explanation for each change in each (online/offline) baseline.
        - If possible, figure out the cause and fix.
          - Discuss with repo or code owner if the fix is a new patch.
          - This may require adding packages to dotnet/source-build-reference-packages.
        - If the baseline change is due to the change in patch version, make sure there's an issue tracking fixing the instability, or start tracking it.

>         - If in doubt, talk to the team. Someone might have already encountered a similar diff in the past.

Author: Added as suggested.

> 1.  - [ ] Add team members as reviewers to the PR. Get acceptance.

Reviewer: Specify who this is.

Author: Added a list.

      - Note: The PR is internal for internal builds and on GitHub for non-internal builds.
      - [Non-Internal] Merge the PR when green and accepted.
      - [Internal] Don't merge the PR.
1.  - [ ] [Internal] Download the artifacts from the centos71_Online_BaseTarball PR validation build onto a machine with Bash.
      - From the "Artifacts" page, navigate to "Tarball centos71_Online_BaseTarball" and download the `base_tarball_*.tar.gz` and `tarball_*-smoke-test-prereqs.tar.gz` files.
      - The smoke-test prereqs are platform-independent but should be taken from this non-portable build, *not* CentOS7 Offline Portable.
      - Can use <https://github.com/dotnet/source-build/blob/release/3.1/scripts/fetch-azdo-artifact-tarball.sh> as a library.
        - E.g. download the tarballs with `download_tarball`, and fix them up with `fix_azdo_tarball` if necessary.
1.  - [ ] [Internal] Rename tarball and smoke-test prereqs to human-friendly names. (Can use `rename_tarball_inner_dir`.)
      - Refer to previous release's email threads for naming pattern.

> 1.  - [ ] [Internal] Upload tarballs and smoke-test prereqs to blob storage.

Reviewer: What blob storage?  Should a source-build team member do this to control access?

Author: I'll add the details for where and where to get the creds anyway, then we can decide whether we want to do it ourselves.  It'll still be useful for new team members.

> 1.  - [ ] Complete manual validation steps

Reviewer: These steps are going to need more explanation so that someone with no or limited SB experience can follow.

Author: Added details and commands to run.

          - [#1631](https://github.com/dotnet/source-build/issues/1631) tracks a fix.
        - Use <https://gist.github.com/omajid/9111a9310d452b3594b770d4ca8a3d87> as a loose baseline.
          - [#1630](https://github.com/dotnet/source-build/issues/1630) tracks a checked-in baseline.

> 1.  - [ ] Coordinate with team and complete manual distributed smoke-testing. Suggested assignment below, but it's flexible.  Send email with direct links to artifacts from green PR.

Author: Yes, changed.

      - [3.1]
        - [ ] RHEL 7 & 8 VM - crummel
        - [ ] Fedora 30 - msimons
        - [ ] CentOS 7 - msimons
        - [ ] CentOS 8 - lbussell
        - [ ] Debian 9 - lbussell

> 1.  - [ ] [Internal] Send the tarball to partners. Include info about how certain we are that this will be the final Microsoft build.

Reviewer: There are a few steps that should be done by a source-build team member.  These should be called out explicitly.  This is one of them.

Author: Added notes where I think this is needed.
